### PR TITLE
clang format

### LINKED ISSUE #544: Propose .clang-format
Try to keep it as close to existing styles as possible. Propose to team for review.

### PR DESCRIPTION
**Pull request by greg7mdp**
Resolves #544 

Initial commit to start the discussion.

### CODE REVIEW

**Comment by greg7mdp** [comment]
This is just an initial test. We can suppress formatting in some cases (like some tests) where the declarations are expanded excessively.

**Comment by greg7mdp** [comment]
You can hide whitespace changes (although not changes in line breaks by adding `?w=1` at the end of the url, or to change it permanently in github in the gui:
image

**Comment by heifner** [comment]
FWIW, I don't think we should reformat the code base. This makes it more difficult for forks to take updates and even maintaining 3.1 & 3.2. Also it adds an extra jump anytime you use git blame to determine why a section of code was modified.

**Comment by greg7mdp** [comment]
> This makes it more difficult for forks to take updates and even maintaining 3.1 & 3.2. 

Not if the forks, as well as 3.1 and 3.1, are also formatted with the same .clang-format.

> Also it adds an extra jump anytime you use git blame to determine why a section of code was modified.

Yes, that is an issue. I feel that in the long run, this issue will diminish, while we will keep enjoying and benefiting from consistently formatted code. But I see how this could be seen as a too big step to take.

**Review by spoonincode — COMMENTED** [review]
I only scrolled through a little bit and dropped some non-exhaustive comments along the way. Reformatting existing code like this seems like a complete non-starter to me as many of the changes included here are clear regressions to the readability, or are otherwise kinda goofy changes only a tool would do (not a human would do), or otherwise miss the nuance of why the code is structured the way it is (the `<<` chains and boost program options stuff being really good examples), or more or less changing things just to change things. Yeah there are some improvements mixed in too.

For the same reasons, enforcing these rules as part of a CI pass seem _extremely_ harsh.

In general I believe that readable code doesn't come about due to strict pedantic adherence to some format rules, but rather with sound design and engineering combined with consistent proper pattern usage. As far as whitespace goes, there is a certain art in attractiveness which tools are oblivious to; as so many examples in this PR illustrate to a tee.

I would encourage us to instead find a set of high level formatting rules to follow which avoid completely jarring contrasts when switching between files. For example,
* indentation is 3 spaces
* braces go on same line

are both rules that have been followed very very well. Admittedly, following on from there may lead to some more contentious rules that are 50/50 in existing code. But it's not clear to me whether, for example, the presence or lack of a space after `if` is really critical to have consistency across all files in the project. (I can certainly appreciate and encourage consistency for these smaller nuances on a function, class, or file level as appropriate)

**Comment by greg7mdp** [comment]
> I only scrolled through a little bit and dropped some non-exhaustive comments along the way. Reformatting existing code like this seems like a complete non-starter to me as many of the changes included here are clear regressions to the readability

Matt, where custom formatting is preferred, and it is true that sometimes it is appropriate, it is easy to bracket with `// clangformat off` and `// clangformat on` comments. I have not spent time doing that in this PR, but if we did agree to proceed, it should be done.

> readable code doesn't come about due to strict pedantic adherence to some format rules

I disagree with that statement, although I would have said: `consistent formatting helps with code readability`.

> enforcing these rules as part of a CI pass seem extremely harsh

Again I disagree, without enforcement, we cannot expect consistent formatting. I don't see what is harsh about it.

File: benchmark/main.cpp [hunk 1]
```diff
@@ -9,17 +9,19 @@ using bpo::options_description;
...
-      ("runs,r", bpo::value<uint32_t>(&num_runs)->default_value(1000), "the number of times running a function during benchmarking")
-      ("help,h", "benchmark functions, and report average, minimum, and maximum execution time in nanoseconds");
+   options_description cli("benchmark command line options");
+   cli.add_options()("feature,f",
+                     bpo::value<std::string>(),
+                     "feature to be benchmarked; if this option is not present, all features are benchmarked.")(
+      "list,l", "list of supported features")("runs,r",
+                                              bpo::value<uint32_t>(&num_runs)->default_value(1000),
+                                              "the number of times running a function during benchmarking")(
+      "help,h", "benchmark functions, and report average, minimum, and maximum execution time in nanoseconds");
```
[the remark this answers is not in the document]
Comment (ScottBailey) [reply benchmark/main.cpp:24]: +1

**Comment by greg7mdp** [comment]
Closing this PR superseded by #652

**Pull request closed without merging**